### PR TITLE
[SYCL][ESIMD] Implement unified memory API - atomic_update(acc,...) with two operands

### PR DESCRIPTION
**Pull request by sarnex**
This implements atomic_update with two operands and accessors. 

We only need four overloads to cover all cases because we can rely on the simd_view to simd implicit conversion operator. The behavior of every possible combination is locked down in a compile time test, and I manually verified the all the combinations are correct and nothing is missed.

I also manually tested on PVC.

### CODE REVIEW

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 1]
```diff
@@ -5862,58 +5919,223 @@ atomic_update(AccessorTy acc, Toffset offset, simd<Tx, N> src0,
...
+        ext::oneapi::experimental::is_property_list_v<PropertyListT>,
+    simd<T, N>>
+atomic_update(AccessorTy acc, simd<Toffset, N> byte_offset, simd<T, N> src0,
+              simd<T, N> src1, PropertyListT props = {}) {
+  simd_mask<N> mask = 1;
+  return atomic_update<Op, T, N>(acc, byte_offset, src0, src1, mask, props);
+}
+
+/// simd<T, N>
+/// atomic_update(AccessorTy acc, simd_view<Toffset, OffsetRegionTy>
```
Comment (turinevgeny) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:6062]: What happens if this overload is removed? For slm/lacc only one overload seems sufficient.

Comment (sarnex) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp:6062]: If I remove that, the following unit test calls begin to fail

```
 auto res_atomic_25 = atomic_update<atomic_op::cmpxchg, int, VL>(
          acc, offsets_view, swap, compare, pred, props_a);

  auto res_atomic_27 = atomic_update<atomic_op::cmpxchg, int, VL>(
        acc, offsets_view, swap, compare_view, pred, props_a);


 auto res_atomic_29 = atomic_update<atomic_op::cmpxchg, int, VL>(
        acc, offsets_view, swap_view, compare, pred, props_a);

auto res_atomic_31 = atomic_update<atomic_op::cmpxchg, int, VL>(
        acc, offsets_view, swap_view, compare_view, pred, props_a);
```

Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp:6062]: Here the offset is not only uint32_t vector as in SLM case.
Using `TOffset` in `simd<TOffset, N> byte_offsets` overload makes the matching work more difficult for C++ FE.

Comment (sarnex) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp:6062]: good point, let me try that

Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp:6062]: Hmm, try what? In my comment, I mean - we most likely need to keep the simd_view overloads for `byte_offsets`

Comment (sarnex) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp:6062]: yeah sorry i was just wondering if that's what the problem was

**Review by turinevgeny — APPROVED** [review]

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 2]
```diff
@@ -5862,58 +5919,223 @@ atomic_update(AccessorTy acc, Toffset offset, simd<Tx, N> src0,
...
+//    Other properties are ignored.
+/// @return A vector of the old values at the memory locations before the
+///   update.
+template <atomic_op Op, typename T, int N, typename Toffset,
+          typename AccessorTy,
+          typename OffsetRegionTy = region1d_t<Toffset, N, 1>,
+          typename PropertyListT =
+              ext::oneapi::experimental::detail::empty_properties_t>
+__ESIMD_API std::enable_if_t<
+    __ESIMD_DNS::get_num_args<Op>() == 2 && !std::is_pointer_v<AccessorTy> &&
```
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:6129]: Gregory is trying to fix all cases like this. Please add the new functions with proper checks.
In this it should be `is_rw_device_accessor_v<AccessorTy>`

**Review by v-klochkov — COMMENTED** [review]
I found some issues - not critical, but deserving a fix within this PR scope.

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 3]
```diff
@@ -3690,6 +3690,63 @@ atomic_update_impl(AccessorTy acc, simd<Toffset, N> byte_offset,
...
+                   simd<T, N> src0, simd<T, N> src1, simd_mask<N> pred) {
+#ifdef __ESIMD_FORCE_STATELESS_MEM
+  return atomic_update_impl<Op, T, N, DS, L1H, L2H>(
+      __ESIMD_DNS::accessorToPointer<T>(acc), byte_offset, src0, src1, pred);
+#else
+  static_assert(std::is_integral_v<Toffset> && sizeof(Toffset) == 4,
+                "Unsupported offset type");
+  check_lsc_vector_size<1>();
+  check_lsc_data_size<T, DS>();
+  check_atomic<Op, T, N, 2, /*IsLSC*/ true>();
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp:3730]: Sorry, I am wrong in my previous comment - check_atomic() function accepts IsLSC and does not check `N` if IsLSC is true.

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 4]
```diff
@@ -5862,58 +5919,223 @@ atomic_update(AccessorTy acc, Toffset offset, simd<Tx, N> src0,
...
+/// simd<T, N>
+/// atomic_update(AccessorTy acc, simd<Toffset, N> byte_offset, simd<T, N> src0,
+///              simd<T, N> src1, simd_mask<N> mask, props = {}); //(acc-au1-2)
+///
+/// simd<T, N>
+/// atomic_update(AccessorTy acc, simd<Toffset, N> byte_offset,
+///               simd<T, N> src0, simd<T, N> src1,
+///               props = {});                                  // (acc-au2-2)
+/// simd<T, N>
+/// atomic_update(AccessorTy acc, simd_view<Toffset, OffsetRegionTy>
```
[the remark this answers is not in the document]
Comment (sarnex) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp]: fixed, thx

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 5]
```diff
@@ -5862,58 +5919,223 @@ atomic_update(AccessorTy acc, Toffset offset, simd<Tx, N> src0,
...
+/// @param mask Operation mask, only locations with non-zero in the
+///   corresponding mask element are updated.
+/// @param props The parameter 'props' specifies the optional compile-time
+///   properties list. Only L1/L2 properties are used.
+//    Other properties are ignored.
+/// @return A vector of the old values at the memory locations before the
+///   update.
+template <atomic_op Op, typename T, int N, typename Toffset,
+          typename AccessorTy,
+          typename OffsetRegionTy = region1d_t<Toffset, N, 1>,
```
[the remark this answers is not in the document]
Comment (sarnex) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp]: removed, thx

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 6]
```diff
@@ -5862,58 +5919,223 @@ atomic_update(AccessorTy acc, Toffset offset, simd<Tx, N> src0,
...
+atomic_update(AccessorTy acc, simd_view<Toffset, OffsetRegionTy> byte_offset,
+              simd<T, N> src0, simd<T, N> src1, simd_mask<N> mask,
+              PropertyListT props = {}) {
+  return atomic_update<Op, T, N>(acc, byte_offset.read(), src0, src1, mask,
+                                 props);
+}
+
+/// simd<T, N>
+/// atomic_update(AccessorTy acc, simd_view<Toffset, OffsetRegionTy>
+/// byte_offset,
```
[the remark this answers is not in the document]
Comment (sarnex) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp]: i redid all the formatting in my latest commit, some clang-format call messed it up and i didn't catch it

File: sycl/include/sycl/ext/intel/esimd/memory.hpp (hunk 2 above)
Comment (sarnex) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp]: merged to get latest changes and added this in latest commit

**Comment by sarnex** [comment]
@intel/llvm-gatekeepers Can we merge this one? The Arc matrix tests haven't run even after 16 hrs  but this PR can't effect that and they [passed](https://github.com/intel/llvm/actions/runs/7065093473/job/19234752598) in an earlier CI run.

**Pull request merged**